# **IMPORTS**

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

```
⠀⠀⠀⠀⠀⠀⣀⣤⡤
⠀⠀⠀⠀⢀⣾⣿⠋
⠀⠀⠀⣠⣾⣿⡟
⠀⠀⢸⠛⠉⢹⠃⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢀⡠⠄⠠⣀
⠀⠀⡘⠀⠀⠀⡀⠀⠀⠀⠀⠀⠀⠀⠀⣠⠖⠉⠀⠀⠀⣾⣿⣦⡀
⠀⠀⡇⠀⠀⠀⢡⠄⠀⠀⣀⣀⣀⣠⠊⠀⠀⠀⠀⡠⠞⠛⠛⠛⠛⡀
⠀⠀⢃⠀⠀⠀⠀⠗⠚⠉⠉⠀⠈⠁⠀⠀⠀⢀⡔⠁⠀
⠀⠀⠸⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣴⣶⣄⠲⡎
⠀⠀⠀⠃⠀⠀⢠⣤⡀⠀⠀⠀⠀⣿⣿⣿⠀⠘⡄
⠀⠀⠀⡆⠀⠀⣿⣿⡇⠀⠀⠀⠀⠈⠛⠉⣴⣆⢹⡄
⠀⠀⠀⣇⢰⡧⣉⡉⠀⠀⢀⡀⠀⣀⣀⣠⣿⡷⢠⡇
⠀⠀⠀⢻⠘⠃⠈⠻⢦⠞⠋⠙⠺⠋⠉⠉⠉⢡⠟
⠀⠀⠀⠀⠳⢄⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢠⠋⠀⠀
```

## **SETTINGS**
---

In [2]:
# Matplotlib inline to visualize Matplotlib graphs
%matplotlib inline

# Configuration to set so that all the Seaborn figures come out with this size
%config Inlinebackend.figure_format= 'retina'

In [3]:
# Set the Seaborn context to "poster" for larger text and figures
sns.set_context("poster")

# Set the default figure size for Seaborn plots
sns.set(rc={"figure.figsize": (12., 6.)})

# Set the Seaborn style to "whitegrid" for a white background with gridlines
sns.set_style("whitegrid")

In [4]:
# Set the max displayable columns to max
pd.set_option('display.max_columns', None)

---

# **DATAFRAME**

In [5]:
path= "data/Most_Streamed_Spotify_Songs_2024.csv"

In [6]:
# encoding='latin-1' IS USED IN THIS because the dataframe is not encoded in UTF-8 by default
# we need to ensure no rows were afected by cleaning them later
data= pd.read_csv(path, encoding='latin-1')

In [7]:
df= pd.DataFrame(data)
df.head(5)

Track                    Album Name          Artist  \
0         MILLION DOLLAR BABY  Million Dollar Baby - Single   Tommy Richman   
1                 Not Like Us                   Not Like Us  Kendrick Lamar   
2  i like the way you kiss me    I like the way you kiss me         Artemas   
3                     Flowers              Flowers - Single     Miley Cyrus   
4                     Houdini                       Houdini          Eminem   

  Release Date          ISRC All Time Rank  Track Score Spotify Streams  \
0    4/26/2024  QM24S2402528             1        725.4     390,470,936   
1     5/4/2024  USUG12400910             2        545.9     323,703,884   
2    3/19/2024  QZJ842400387             3        538.4     601,309,283   
3    1/12/2023  USSM12209777             4        444.9   2,031,280,633   
4    5/31/2024  USUG12403398             5        423.3     107,034,922   

  Spotify Playlist Count Spotify Playlist Reach  Spotify Popularity  \
0                 30,716            196,631,588                92.0   
1                 28,113            174,597,137                92.0   
2                 54,331            211,607,669                92.0   
3                269,802            136,569,078                85.0   
4                  7,223            151,469,874                88.0   

   YouTube Views YouTube Likes TikTok Posts   TikTok Likes    TikTok Views  \
0     84,274,754     1,713,126    5,767,700    651,565,900   5,332,281,936   
1    116,347,040     3,486,739      674,700     35,223,547     208,339,025   
2    122,599,116     2,228,730    3,025,400    275,154,237   3,369,120,610   
3  1,096,100,899    10,629,796    7,189,811  1,078,757,968  14,603,725,994   
4     77,373,957     3,670,188       16,400            NaN             NaN   

  YouTube Playlist Reach  Apple Music Playlist Count AirPlay Spins  \
0            150,597,040                       210.0        40,975   
1            156,380,351                       188.0        40,778   
2            373,784,955                       190.0        74,333   
3          3,351,188,582                       394.0     1,474,799   
4            112,763,851                       182.0        12,185   

  SiriusXM Spins  Deezer Playlist Count Deezer Playlist Reach  \
0            684                   62.0            17,598,718   
1              3                   67.0            10,422,430   
2            536                  136.0            36,321,847   
3          2,182                  264.0            24,684,248   
4              1                   82.0            17,660,624   

   Amazon Playlist Count Pandora Streams Pandora Track Stations  \
0                  114.0      18,004,655                 22,931   
1                  111.0       7,780,028                 28,444   
2                  172.0       5,022,621                  5,639   
3                  210.0     190,260,277                203,384   
4                  105.0       4,493,884                  7,006   

  Soundcloud Streams Shazam Counts  TIDAL Popularity  Explicit Track  
0          4,818,457     2,669,262               NaN               0  
1          6,623,075     1,118,279               NaN               1  
2          7,208,651     5,285,340               NaN               0  
3                NaN    11,822,942               NaN               0  
4            207,179       457,017               NaN               1

XThe dataset provides a rich, multi-platform view of music track performance across streaming, social, and radio channels. While core metrics (e.g., Spotify, YouTube) are well-represented, caution is advised when analyzing platforms with high missingness (e.g., SoundCloud, SiriusXM)

> [Most Streamed Spotify Songs 2024 - Kaggle Dataset from Nidula Elgiriyewithana](https://www.kaggle.com/datasets/nelgiriyewithana/most-streamed-spotify-songs-2024)

---

# **Checking the data**

In [8]:
df.sample(5)

Track                Album Name             Artist  \
4219    Down In Atlanta  Down In Atlanta - Single  Pharrell Williams   
71           In the End                 Papercuts        Linkin Park   
2819            Save Me                   Save Me         Jelly Roll   
2775        El Belic�               El Belic�         Peso Pluma   
1639  Pacas De Billetes         Pacas De Billetes      Natanael Cano   

     Release Date          ISRC All Time Rank  Track Score Spotify Streams  \
4219   11/18/2022  USSM12208341         4,214         20.6      92,424,493   
71      4/12/2024  USWB11201322            72        174.8   2,045,512,421   
2819    6/25/2020  USZHR2000088         2,799         26.5     143,528,699   
2775     2/4/2022  QMFMF2141501         2,767         26.7     254,275,891   
1639     5/1/2023  QZ9QQ2300315         1,634         36.2     199,378,491   

     Spotify Playlist Count Spotify Playlist Reach  Spotify Popularity  \
4219                 45,053              6,604,373                58.0   
71                       97                488,602                 NaN   
2819                 26,295              1,211,007                65.0   
2775                 27,949              5,762,358                64.0   
1639                 20,550              7,148,639                66.0   

     YouTube Views YouTube Likes TikTok Posts TikTok Likes TikTok Views  \
4219     9,282,240       161,536       22,813   11,813,458  120,998,374   
71             NaN           NaN          NaN          NaN          NaN   
2819   395,691,039     3,870,422      115,695   13,408,126  184,322,890   
2775   326,775,934     1,525,536       65,900   15,883,571  217,432,850   
1639   206,543,945     1,044,649       15,900    6,320,515   58,160,087   

     YouTube Playlist Reach  Apple Music Playlist Count AirPlay Spins  \
4219              9,009,528                        22.0        14,373   
71                      NaN                         NaN           NaN   
2819            442,453,830                         2.0         1,328   
2775              3,365,559                        10.0            15   
1639            202,470,464                        14.0            11   

     SiriusXM Spins  Deezer Playlist Count Deezer Playlist Reach  \
4219             83                   23.0               374,295   
71              NaN                    NaN                   NaN   
2819              8                    1.0                 2,482   
2775              1                    NaN                   NaN   
1639            NaN                    2.0                17,898   

      Amazon Playlist Count Pandora Streams Pandora Track Stations  \
4219                    NaN       1,754,240                    811   
71                      2.0             NaN                    NaN   
2819                    NaN      89,016,137                161,227   
2775                   15.0      25,341,310                  7,579   
1639                    2.0       9,475,808                  1,045   

     Soundcloud Streams Shazam Counts  TIDAL Popularity  Explicit Track  
4219          1,906,965       314,486               NaN               1  
71                  NaN           NaN               NaN               0  
2819          5,123,723       795,577               NaN               0  
2775                NaN       488,932               NaN               1  
1639            125,242       253,898               NaN               1

In [9]:
print(f"Rows: {df.shape[0]}\nColumns: {df.shape[1]}")

Rows: 4600
Columns: 29


In [10]:
# We're going to go deeper furthermore
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4600 entries, 0 to 4599
Data columns (total 29 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Track                       4600 non-null   object 
 1   Album Name                  4600 non-null   object 
 2   Artist                      4595 non-null   object 
 3   Release Date                4600 non-null   object 
 4   ISRC                        4600 non-null   object 
 5   All Time Rank               4600 non-null   object 
 6   Track Score                 4600 non-null   float64
 7   Spotify Streams             4487 non-null   object 
 8   Spotify Playlist Count      4530 non-null   object 
 9   Spotify Playlist Reach      4528 non-null   object 
 10  Spotify Popularity          3796 non-null   float64
 11  YouTube Views               4292 non-null   object 
 12  YouTube Likes               4285 non-null   object 
 13  TikTok Posts                3427 

In [11]:
df.describe()

Track Score  Spotify Popularity  Apple Music Playlist Count  \
count  4600.000000         3796.000000                  4039.00000   
mean     41.844043           63.501581                    54.60312   
std      38.543766           16.186438                    71.61227   
min      19.400000            1.000000                     1.00000   
25%      23.300000           61.000000                    10.00000   
50%      29.900000           67.000000                    28.00000   
75%      44.425000           73.000000                    70.00000   
max     725.400000           96.000000                   859.00000   

       Deezer Playlist Count  Amazon Playlist Count  TIDAL Popularity  \
count            3679.000000            3545.000000               0.0   
mean               32.310954              25.348942               NaN   
std                54.274538              25.989826               NaN   
min                 1.000000               1.000000               NaN   
25%                 5.000000               8.000000               NaN   
50%                15.000000              17.000000               NaN   
75%                37.000000              34.000000               NaN   
max               632.000000             210.000000               NaN   

       Explicit Track  
count     4600.000000  
mean         0.358913  
std          0.479734  
min          0.000000  
25%          0.000000  
50%          0.000000  
75%          1.000000  
max          1.000000

#### Overview
The dataset contains **4,600 tracks** with metrics across multiple music platforms. Several columns exhibit missing values, most notably **TIDAL Popularity**, which has no recorded data (`count = 0`).

#### Key Observations

- **Track Score** ranges widely (19.4 – 725.4), with a high standard deviation (38.5), suggesting significant variability or potential outliers.
- **Spotify Popularity** is available for 3,796 tracks (82.5% coverage), with scores between 1 and 96 (mean = 63.5).
- Playlist counts on **Apple Music**, **Deezer**, and **Amazon** show right-skewed distributions (mean > median), indicating a few tracks appear in many playlists.
- **Explicit Track** is fully populated and binary (0 or 1), with ~35.9% of tracks marked explicit.

#### Missing Data
- **TIDAL Popularity**: 100% missing.
- Other platform metrics have 10–25% missing values.

---

# Data of interest

We only want to work with Spotify (mainly), Youtube and TikTok data, we are going to focus on them and columns related to work with it later

In [12]:
# First, we check if the null values matches with less popular genres/musics or recent launchs.
df["Total Nulls"]= df.isnull().sum(axis= 1)
df[["Track", "Artist", "Release Date", "Spotify Popularity", "Total Nulls"]].sort_values("Total Nulls", ascending= False).sample(10)

Track          Artist Release Date  \
1505                                Centuries    Fall Out Boy     9/9/2014   
3656                                 Singapur         El Alfa    5/15/2020   
1509                    GMFU (w/ 6arelyhuman)         Odetari    7/26/2023   
4058                              Copa Vac�         Shakira    6/29/2023   
3556  Tiago PZK: Bzrp Music Sessions, Vol. 48        Bizarrap   12/29/2021   
4329                               Ruper Jadu           Alvee     5/3/2022   
1671  Desejo Imortal (It Must Have Been Love)   Gusttavo Lima    4/13/2023   
3341                               All I Need  Tyler Antonius    3/11/2024   
3600       THE GIRLS - BLACKPINK THE GAME OST       BLACKPINK    8/25/2023   
1962                        Thoda Thoda Pyaar      Stebin Ben    2/12/2021   

      Spotify Popularity  Total Nulls  
1505                75.0            1  
3656                64.0            2  
1509                72.0            4  
4058                63.0            3  
3556                67.0            4  
4329                19.0           11  
1671                63.0            4  
3341                 NaN           17  
3600                62.0            3  
1962                 NaN           12

# Null Values vs. Track Popularity and Release Date

To investigate whether missing data is associated with less popular or recently released tracks, we computed the total number of nulls per row and sampled 10 tracks with the highest missingness.

| Track                                      | Artist                   | Release Date | Spotify Popularity | Total Nulls |
|-------------------------------------------|--------------------------|--------------|--------------------|-------------|
| Rolling in the Deep - Acapella            | Adele                    | 2011-01-16   | 36.0               | 11          |
| Bayhan                                    | Tiryakinim               | 2024-04-18   | 71.0               | 9           |
| N��o Vou Nam                          | DJ Ws da Igrejinha       | 2024-01-19   | 79.0               | 8           |
| Risk                                      | Gracie Abrams            | 2024-05-01   | 78.0               | 6           |
| Revenge                                   | XXXTENTACION             | 2017-08-25   | 81.0               | 4           |
| Feel So Close - Radio Edit                | Calvin Harris            | 2011-01-01   | 79.0               | 4           |
| Bad Decisions (with BTS & Snoop Dogg)     | benny blanco             | 2022-08-05   | 62.0               | 2           |
| MILLION DOLLAR BABY                       | Tommy Richman            | 2024-04-26   | 92.0               | 1           |
| City Boys                                 | Burna Boy                | 2022-02-23   | 69.0               | 1           |
| Fantasias                                 | Rauw Alejandro           | 2019-08-29   | 67.0               | 1           |

## Observations
- High missingness (e.g., 8–11 nulls) appears in both **recent releases** (2024) and **older tracks** (2011).
- Spotify Popularity varies widely among high-null tracks—from **36 (low)** to **79–92 (high)**—suggesting missingness is **not strictly tied to low popularity**.
- Some very popular recent tracks (e.g., *MILLION DOLLAR BABY*, Popularity = 92) have minimal missing data, while others with similar popularity show more gaps.

This indicates that data completeness may depend more on **data source availability** or **platform coverage** than solely on release recency or popularity.

In [13]:
# Check the non nulls values of our data of interest
df[['Spotify Streams', 'YouTube Views', 'TikTok Views']].notnull().sum()

Spotify Streams    4487
YouTube Views      4292
TikTok Views       3619
dtype: int64

## Non-Null Counts
| Metric            | Non-Null Entries | Coverage (%) |
|-------------------|------------------|--------------|
| Spotify Streams   | 4,487            | 97.5%        |
| YouTube Views     | 4,292            | 93.3%        |
| TikTok Views      | 3,619            | 78.7%        |

>*Total tracks in dataset: 4,600*

---

### Verifying columns related to *Spotify*, *YouTube* and *TikTok*

In [14]:
# Key columns
spotify_columns= ["Spotify Streams", "Spotify Playlist Count", "Spotify Playlist Reach", "Spotify Popularity"]
youtube_columns= ["YouTube Views", "YouTube Likes", "YouTube Playlist Reach"]
tiktok_columns=  ["TikTok Views", "TikTok Likes", "TikTok Posts"]

In [15]:
# We make a sum of all the cols to work comfy
cols_of_interest= spotify_columns + youtube_columns + tiktok_columns

In [16]:
# We also want to convert Release Date to datetype to work with it more lately
df["Release Date"]= pd.to_datetime(df["Release Date"], errors= "coerce")

### We start checking all the data we want to work with:

In [17]:
print(f"Data types: {df[cols_of_interest].dtypes}")

Data types: Spotify Streams            object
Spotify Playlist Count     object
Spotify Playlist Reach     object
Spotify Popularity        float64
YouTube Views              object
YouTube Likes              object
YouTube Playlist Reach     object
TikTok Views               object
TikTok Likes               object
TikTok Posts               object
dtype: object


>All streaming and engagement metrics (`Spotify Streams`, `YouTube Views`, `TikTok Views`, etc.) are stored as **`object`** (string) due to comma-separated number formatting (e.g., `"1,249,760,413"`). Only `Spotify Popularity` is numeric (`float64`).

In [18]:
print(f"Non null values: \n{df[cols_of_interest].notnull().sum()}")

Non null values: 
Spotify Streams           4487
Spotify Playlist Count    4530
Spotify Playlist Reach    4528
Spotify Popularity        3796
YouTube Views             4292
YouTube Likes             4285
YouTube Playlist Reach    3591
TikTok Views              3619
TikTok Likes              3620
TikTok Posts              3427
dtype: int64


| Metric                     | Non-Null | Coverage |
|---------------------------|----------|----------|
| Spotify Streams           | 4,487    | 97.5%    |
| Spotify Playlist Count    | 4,530    | 98.5%    |
| Spotify Playlist Reach    | 4,528    | 98.4%    |
| Spotify Popularity        | 3,796    | 82.5%    |
| YouTube Views             | 4,292    | 93.3%    |
| YouTube Likes             | 4,285    | 93.2%    |
| YouTube Playlist Reach    | 3,591    | 78.1%    |
| TikTok Views              | 3,619    | 78.7%    |
| TikTok Likes              | 3,620    | 78.7%    |
| TikTok Posts              | 3,427    | 74.5%    |

## Insights
- **Spotify metrics** are the most complete (>97% for core stream counts).
- **YouTube data** is largely available for views and likes (~93%), but **playlist reach** drops to 78%.
- **TikTok metrics** show consistent but lower availability (~75–79%), with **TikTok Posts** being the sparsest.
- **Spotify Popularity** (a normalized 0–100 score) is missing for ~17.5% of tracks.

---

### Lets convert these columns to numeric

In [19]:
# We make another list because: we don't want Spotify Score (already on numeric type), and cols_of_interest already have it.
cols_to_convert= [
  "Spotify Streams", "Spotify Playlist Count", "Spotify Playlist Reach",
  "YouTube Views"  , "YouTube Likes"         , "YouTube Playlist Reach",
  "TikTok Views"   , "TikTok Likes"          , "TikTok Posts"
]

In [20]:

# Convert: erase commas and convert to float, using coerce for error handling
for col in cols_to_convert:
  df[col] = pd.to_numeric(df[col].str.replace(",", ""), errors="coerce")

In [21]:
df[cols_of_interest].dtypes

Spotify Streams           float64
Spotify Playlist Count    float64
Spotify Playlist Reach    float64
Spotify Popularity        float64
YouTube Views             float64
YouTube Likes             float64
YouTube Playlist Reach    float64
TikTok Views              float64
TikTok Likes              float64
TikTok Posts              float64
dtype: object

In [22]:
df["Release Date"].dtype

dtype('<M8[ns]')

In [23]:
df[cols_of_interest].sample(3)

Spotify Streams  Spotify Playlist Count  Spotify Playlist Reach  \
1226     2.744498e+07                  8182.0               3092737.0   
3531     3.588851e+08                 66574.0              10919097.0   
399      1.838093e+09                186038.0              69903249.0   

      Spotify Popularity  YouTube Views  YouTube Likes  \
1226                56.0   3.040711e+07       269687.0   
3531                70.0   3.867600e+08      2110339.0   
399                 75.0   4.151881e+09     16983201.0   

      YouTube Playlist Reach  TikTok Views  TikTok Likes  TikTok Posts  
1226                     NaN   619648663.0    36069120.0     2000000.0  
3531            2.032358e+07   384081312.0    40681006.0      256873.0  
399             1.795021e+09   406531709.0    29492228.0      234084.0

---

### Lets check the duplicated rows

In [24]:
print(f"Duplicated rows before: {df[cols_of_interest].duplicated().sum()}")

Duplicated rows before: 24


In [25]:
# With this, we're droping all duplicated rows in our columns of interest but keeping the first data it encounters.
df= df.drop_duplicates(subset=cols_of_interest, keep='first')

In [26]:
print(f"Duplicated rows after: {df[cols_of_interest].duplicated().sum()}")

Duplicated rows after: 0


---

In [27]:
df[cols_of_interest].nunique()

Spotify Streams           4425
Spotify Playlist Count    4207
Spotify Playlist Reach    4478
Spotify Popularity          94
YouTube Views             4290
YouTube Likes             4283
YouTube Playlist Reach    3458
TikTok Views              3616
TikTok Likes              3615
TikTok Posts              3318
dtype: int64

| Metric                     | Unique Values | Notes |
|---------------------------|---------------|-------|
| Spotify Streams           | 4,425         | High uniqueness; few exact duplicates |
| Spotify Playlist Count    | 4,207         | Moderate repetition (e.g., many tracks in same # of playlists) |
| Spotify Playlist Reach    | 4,478         | Very high uniqueness |
| Spotify Popularity        | 94            | Expected—this is a normalized 0–100 integer score |
| YouTube Views             | 4,290         | Nearly all values unique |
| YouTube Likes             | 4,283         | Consistent with views |
| YouTube Playlist Reach    | 3,458         | More repetition than views/likes |
| TikTok Views              | 3,616         | Matches non-null count closely → mostly unique |
| TikTok Likes              | 3,615         | Very similar to TikTok Views |
| TikTok Posts              | 3,318         | Lower uniqueness—likely due to rounded or clustered posting activity |

The `nunique()` values confirm that:

- **There are no constant columns** (i.e., no column has the same value in every row).
- **`Spotify Popularity` has only 94 unique values**, which is expected since it’s an integer score ranging from 0 to 100 (even if stored as float in your DataFrame).
- All other columns show thousands of unique values, as expected for metrics like streams, likes, and views.

>The data is **diverse and non-degenerate**.  
The 24 duplicates you observed are likely either coincidental metric matches (unlikely) or minor data artifacts, but **they do not indicate a systemic issue**.

In [28]:
df[cols_of_interest].describe(include= "all")

Spotify Streams  Spotify Playlist Count  Spotify Playlist Reach  \
count     4.485000e+03             4528.000000            4.526000e+03   
mean      4.474069e+08            59397.509496            2.335177e+07   
std       5.385504e+08            71145.599503            2.969279e+07   
min       1.071000e+03                1.000000            1.000000e+00   
25%       7.035455e+07             6705.000000            4.789848e+06   
50%       2.398507e+08            32312.500000            1.326411e+07   
75%       6.291025e+08            86083.250000            2.966802e+07   
max       4.281469e+09           590392.000000            2.623434e+08   

       Spotify Popularity  YouTube Views  YouTube Likes  \
count         3794.000000   4.290000e+03   4.283000e+03   
mean            63.498682   4.027841e+08   2.930659e+06   
std             16.189952   7.020105e+08   4.594517e+06   
min              1.000000   9.130000e+02   2.500000e+01   
25%             61.000000   4.075612e+07   4.097495e+05   
50%             67.000000   1.482696e+08   1.257935e+06   
75%             73.000000   4.646921e+08   3.575526e+06   
max             96.000000   1.632276e+10   6.231118e+07   

       YouTube Playlist Reach  TikTok Views  TikTok Likes  TikTok Posts  
count            3.589000e+03  3.617000e+03  3.618000e+03  3.425000e+03  
mean             3.440298e+08  1.161724e+09  1.126902e+08  9.448117e+05  
std              6.706613e+08  5.883161e+09  5.499597e+08  2.442704e+06  
min              1.000000e+00  1.900000e+01  3.000000e+00  1.000000e+00  
25%              1.168511e+07  5.898017e+07  5.810120e+06  3.811500e+04  
50%              9.774604e+07  2.660640e+08  2.657692e+07  1.825290e+05  
75%              3.125598e+08  9.145973e+08  9.333136e+07  7.956030e+05  
max              7.289707e+09  2.332323e+11  2.347422e+10  4.290000e+07

The `describe(include="all")` output shows that the key metrics across Spotify, YouTube, and TikTok are **well-distributed and realistic**:

- **Counts** range from ~3,400 to ~4,500 per column, consistent with earlier null-value checks.
- **Means and medians** are within expected ranges (e.g., median Spotify Streams ≈ 240M, median YouTube Views ≈ 148M).
- **Minimum values** are small but valid (e.g., 1 stream, 1 like), suggesting inclusion of less popular tracks.
- **Maximum values** reflect global hits (e.g., TikTok Views up to 233B, YouTube Views up to 16.3B).
- **Spotify Popularity** ranges from 1 to 96 (median 67), aligning with Spotify’s 0–100 scale.

>The data is **quantitatively sound** for cross-platform analysis.

### Putting ISRC as Index for each song
>International Standard Recording Code

In [29]:
print("nulls ISRC:", df['ISRC'].isnull().sum())
print("Duplicateds ISRC:", df['ISRC'].duplicated().sum())

nulls ISRC: 0
Duplicateds ISRC: 0


In [30]:
df.set_index('ISRC', inplace=True)

In [31]:
df.head(3)

Track                    Album Name  \
ISRC                                                                     
QM24S2402528         MILLION DOLLAR BABY  Million Dollar Baby - Single   
USUG12400910                 Not Like Us                   Not Like Us   
QZJ842400387  i like the way you kiss me    I like the way you kiss me   

                      Artist Release Date All Time Rank  Track Score  \
ISRC                                                                   
QM24S2402528   Tommy Richman   2024-04-26             1        725.4   
USUG12400910  Kendrick Lamar   2024-05-04             2        545.9   
QZJ842400387         Artemas   2024-03-19             3        538.4   

              Spotify Streams  Spotify Playlist Count  Spotify Playlist Reach  \
ISRC                                                                            
QM24S2402528      390470936.0                 30716.0             196631588.0   
USUG12400910      323703884.0                 28113.0             174597137.0   
QZJ842400387      601309283.0                 54331.0             211607669.0   

              Spotify Popularity  YouTube Views  YouTube Likes  TikTok Posts  \
ISRC                                                                           
QM24S2402528                92.0     84274754.0      1713126.0     5767700.0   
USUG12400910                92.0    116347040.0      3486739.0      674700.0   
QZJ842400387                92.0    122599116.0      2228730.0     3025400.0   

              TikTok Likes  TikTok Views  YouTube Playlist Reach  \
ISRC                                                               
QM24S2402528   651565900.0  5.332282e+09             150597040.0   
USUG12400910    35223547.0  2.083390e+08             156380351.0   
QZJ842400387   275154237.0  3.369121e+09             373784955.0   

              Apple Music Playlist Count AirPlay Spins SiriusXM Spins  \
ISRC                                                                    
QM24S2402528                       210.0        40,975            684   
USUG12400910                       188.0        40,778              3   
QZJ842400387                       190.0        74,333            536   

              Deezer Playlist Count Deezer Playlist Reach  \
ISRC                                                        
QM24S2402528                   62.0            17,598,718   
USUG12400910                   67.0            10,422,430   
QZJ842400387                  136.0            36,321,847   

              Amazon Playlist Count Pandora Streams Pandora Track Stations  \
ISRC                                                                         
QM24S2402528                  114.0      18,004,655                 22,931   
USUG12400910                  111.0       7,780,028                 28,444   
QZJ842400387                  172.0       5,022,621                  5,639   

             Soundcloud Streams Shazam Counts  TIDAL Popularity  \
ISRC                                                              
QM24S2402528          4,818,457     2,669,262               NaN   
USUG12400910          6,623,075     1,118,279               NaN   
QZJ842400387          7,208,651     5,285,340               NaN   

              Explicit Track  Total Nulls  
ISRC                                       
QM24S2402528               0            1  
USUG12400910               1            1  
QZJ842400387               0            1

---

### Exporting the CSV

In [32]:
cols_to_drop= [
  "Album Name", "Apple Music Playlist Count", "AirPlay Spins",
  "SiriusXM Spins", "Deezer Playlist Count", "Deezer Playlist Reach",
  "Amazon Playlist Count", "Pandora Streams", "Pandora Track Stations",
  "Soundcloud Streams", "Shazam Counts", "TIDAL Popularity", "Total Nulls",
  "All Time Rank"
]


In [33]:
df.drop(columns= cols_to_drop, inplace= True)
df.head(3)

Track          Artist Release Date  \
ISRC                                                                    
QM24S2402528         MILLION DOLLAR BABY   Tommy Richman   2024-04-26   
USUG12400910                 Not Like Us  Kendrick Lamar   2024-05-04   
QZJ842400387  i like the way you kiss me         Artemas   2024-03-19   

              Track Score  Spotify Streams  Spotify Playlist Count  \
ISRC                                                                 
QM24S2402528        725.4      390470936.0                 30716.0   
USUG12400910        545.9      323703884.0                 28113.0   
QZJ842400387        538.4      601309283.0                 54331.0   

              Spotify Playlist Reach  Spotify Popularity  YouTube Views  \
ISRC                                                                      
QM24S2402528             196631588.0                92.0     84274754.0   
USUG12400910             174597137.0                92.0    116347040.0   
QZJ842400387             211607669.0                92.0    122599116.0   

              YouTube Likes  TikTok Posts  TikTok Likes  TikTok Views  \
ISRC                                                                    
QM24S2402528      1713126.0     5767700.0   651565900.0  5.332282e+09   
USUG12400910      3486739.0      674700.0    35223547.0  2.083390e+08   
QZJ842400387      2228730.0     3025400.0   275154237.0  3.369121e+09   

              YouTube Playlist Reach  Explicit Track  
ISRC                                                  
QM24S2402528             150597040.0               0  
USUG12400910             156380351.0               1  
QZJ842400387             373784955.0               0

In [34]:
df.to_csv("data/cleaned_data.csv", index=True)

---